src/high_level/hash.rs

```diff
@@ -67,6 +79,27 @@ pub fn digest(data: &[u8]) -> Result<Digest, UnknownCryptoError> {
     blake2b::Hasher::Blake2b256.digest(data)
 }
 
+/// Hash data from `impl Read` using BLAKE2b-256.
+///
+/// See the [module-level docs](crate::hash) for an example of how to use this function.
+/// Internally calls `std::io::copy()` to move data from the reader into the Blake2b writer.
+/// The `std::io::copy` function buffers reads, so passing in a `BufReader` may be unnecessary.
+///
+/// For lower-level control over reads, writes, buffer sizes, *etc.*, consider using the
+/// [`Blake2b`](crate::hazardous::hash::blake2::blake2b::Blake2b) type and its
+/// [`Write`](std::io::Write) implementation directly. See the Blake2b type's source code
+/// for an example.
+///
+/// Note that if an error is returned, data may still have been consumed
+/// from the given reader.
+#[cfg(feature = "safe_api")]
+#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+pub fn digest_from_reader(mut reader: impl std::io::Read) -> Result<Digest, UnknownCryptoError> {
```

Comment:
I'm not well-versed in `Read`-based API's, but is it normal to take ownership of the reader as is done here? The reason I'm asking is that, it is only passed as a reference to `std::io::copy()` and I don't immediately see a need to move the reader.

Reply:
So I think that users can pass *any* type that implements Read, and that includes a `&mut R` where `R` is a reader. That's how I've done it in other projects. In that case the unique reference will be passed by value and "consumed", but the original object will be left intact.

Reply:
> So I think that users can pass any type that implements Read, and that includes a &mut R where R is a reader. That's how I've done it in other projects. In that case the unique reference will be passed by value and "consumed", but the original object will be left intact.

I didn't know you could do this, but just tested it and seems to work exactly as intended. Awesome!